### ISSUE TITLE
Unable to login: NullPointerException

### ISSUE THREAD

**Issue by wetneb** [Wikidata/Wikidata-Toolkit#271]
Logging in with this library can fail with a NullPointerException.

Here is my code:
<pre>
import org.wikidata.wdtk.wikibaseapi.ApiConnection;
import org.wikidata.wdtk.wikibaseapi.LoginFailedException;

public class WikidataTest {
	public static void main(String [] args) {
		ApiConnection conn = ApiConnection.getWikidataApiConnection();
		try {
			conn.login("this is a valid username", "corresponding password");
		} catch (LoginFailedException e) {
			e.printStackTrace();
		}
	}
}
</pre>

I obtain the following trace:
<pre>
Exception in thread "main" java.lang.NullPointerException
	at java.net.URLEncoder.encode(URLEncoder.java:205)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.getQueryString(ApiConnection.java:602)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.sendRequest(ApiConnection.java:330)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.confirmLogin(ApiConnection.java:478)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.login(ApiConnection.java:258)
	at WikidataTest.main(WikidataTest.java:9)
</pre>

**Comment by wetneb** [comment]
So, after hours (sigh) of investigation, it looks like the POST requests made by the toolkit are not well-formed enough. Pointing the toolkit to a local server that replays the requests to the Wikidata API fixes the bug. It looks like this problem could be solved by using a higher-level HTTP Java library which would take care of the heavy-lifting in a clean way. That would add a dependency to the toolkit, but hey… it's currently unusable to perform any edits!

**Comment by PolyglotOpenstreetmap** [comment]
Could it be that this fix is not in wikidata-toolkit-with-dependencies-0.7.0.jar, available for download? I'm using the jar file from Jython, but I can't log in.

**Comment by wetneb** [comment]
@PolyglotOpenstreetmap yes this fix has not been released yet… I guess it is quite urgent to make a release. @Tpt let me know if I can help with that!

**Comment by PolyglotOpenstreetmap** [comment]
I don't understand why this issue was closed. I downloaded the jar again and it's still giving the same error message.

**Comment by wetneb** [comment]
@PolyglotOpenstreetmap have you tried building the toolkit from source? downloading released jars again will not solve your problem.

**Comment by PolyglotOpenstreetmap** [comment]
Had I been able to build it myself, I would probably not even have wasted a few hours trying to figure out what I did wrong. Java is not my cup of tea though, so I code in Python, using Jython as a go-between. It worked just fine by addint the jar file to the classpath in the past.

What I didn't realise is that the ticket wasn't reopened automatically when I wrote a message. Then I figured out it's not even possible to do so with github, hence the new ticket.

I'll now go back to waiting patiently for a new release.

Cheers,

Jo
